src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c

```diff
@@ -10,9 +10,241 @@
+        tl_debug(lib, "failed to get ipoib interface for devname %s", ctx->devname);
+        status = UCC_ERR_NOT_SUPPORTED;
+        if (!user_provided_ib) {
+            ucc_free(ctx->devname);
+        }
+        goto error;
+    }
+
+    is_ipv4     = (ip_oib_addr.ss_family == AF_INET) ? 1 : 0;
+    in_src_addr = (struct sockaddr_in*)&ip_oib_addr;
+
+    dst = inet_ntop((is_ipv4) ? AF_INET : AF_INET6,
+                    &in_src_addr->sin_addr, addrstr, sizeof(addrstr) - 1);
+    if (NULL == dst) {
+        tl_error(lib, "inet_ntop failed");
+        status = UCC_ERR_NO_RESOURCE;
+        goto error;
+    }
+
+    tl_debug(ctx->lib, "devname %s, ipoib %s", ctx->devname, addrstr);
+
+    ctx->channel = rdma_create_event_channel();
+    if (!ctx->channel) {
+        tl_error(lib, "rdma_create_event_channel failed, errno %d", errno);
+        status = UCC_ERR_NO_RESOURCE;
+        goto error;
+    }
+
+    memset(&dst_addr, 0, sizeof(struct sockaddr_storage));
+    dst_addr.ss_family = is_ipv4 ? AF_INET : AF_INET6;
+    if (rdma_create_id(ctx->channel, &ctx->id, NULL, RDMA_PS_UDP)) {
+        tl_debug(lib, "failed to create rdma id, errno %d", errno);
+        status = UCC_ERR_NOT_SUPPORTED;
+        goto error;
+    }
+
+    if (0 != rdma_resolve_addr(ctx->id, (struct sockaddr *)&ip_oib_addr,
+                               (struct sockaddr *) &dst_addr, 1000)) {
+        tl_debug(lib, "failed to resolve rdma addr, errno %d", errno);
+        status = UCC_ERR_NOT_SUPPORTED;
+        goto error;
+    }
+
+    if (rdma_get_cm_event(ctx->channel, &revent) < 0) {
+        tl_error(lib, "failed to get cm event, errno %d", errno);
+        status = UCC_ERR_NO_RESOURCE;
+        goto error;
+    } else if (revent->event != RDMA_CM_EVENT_ADDR_RESOLVED) {
+        tl_error(lib, "cm event is not resolved");
+        if (rdma_ack_cm_event(revent) < 0) {
+            tl_error(lib, "rdma_ack_cm_event failed");
+        }
+        status = UCC_ERR_NO_RESOURCE;
+        goto error;
+    }
+
+    if (rdma_ack_cm_event(revent) < 0) {
+        tl_error(lib, "rdma_ack_cm_event failed");
+        status = UCC_ERR_NO_RESOURCE;
+        goto error;
+    }
+
+    ctx->ctx = ctx->id->verbs;
+    ctx->pd  = ibv_alloc_pd(ctx->ctx);
+    if (!ctx->pd) {
+        tl_error(lib, "failed to allocate pd");
+        status = UCC_ERR_NO_RESOURCE;
+        goto error;
+    }
+
+    ib = strdup(ctx->devname);
+    ucc_string_split(ib, ":", 2, &ib_name, &port);
+    ucc_free(ib);
+    ctx->ib_port = atoi(port);
+
+    /* Determine MTU */
+    if (ibv_query_port(ctx->ctx, ctx->ib_port, &port_attr)) {
+        tl_error(lib, "couldn't query port in ctx create, errno %d", errno);
+        status = UCC_ERR_NO_RESOURCE;
+        goto error;
+    }
+
+
+    for (i = 0; i < UCC_TL_MLX5_MCAST_MAX_MTU_COUNT; i++) {
+        if (mtu_lookup[i][1] == port_attr.max_mtu) {
+            max_mtu = mtu_lookup[i][0];
+            break;
+        }
+    }
```

Comment:
why do you need for loop? max_mtu is enum just simple max_mtu = mtu_lookup[ptrt_attr.max_mtu]

Reply:
what if in the future the order of enum list changes and we forget to update that mtu_lookup array.  I wanted us to be on the safe side, otherwise I could simply use 2^(7+enum) and get the mtu

Reply:
I don't think rdma core will ever change this enum and break ABI. Also with such change how your approach will work?

Reply:
enum starts from 1. So, for like IBV_MTU_256 == 1, to achieve 256, I could do pow(2, IBV_MTU_256+7) and get 2^8 = 256 out of it. Should I still change this for loop? leave it as is?

Reply:
I don't mind using pow 2 or array lookup or even switch over enum, just avoid duplication of same code.

Reply:
alright, I just merged two for loops into a single one to avoid duplication.